### PR TITLE
Add document:update

### PR DESCRIPTION
**Pull request by Yoann-Abbes**
## What does this PR do ?

This PR implements the `document:update` method with its unit tests.

### How should this be manually tested?

Clone this branch and run unit tests
`./gradlew test`

When it succeed, compile it

`./gradlew jar`

Initiate another java project by adding the compiled SDK as a dependency.

Then, run Kuzzle, create an index `nyc-open-data` and a `yellow-taxi` collection in the admin console. Create a document with id `some-id` and content:
```
{
        "name" : "random"
}
```
Finally, run this code

```java
import io.kuzzle.sdk.*;
import io.kuzzle.sdk.Options.KuzzleOptions;
import io.kuzzle.sdk.Options.Protocol.WebSocketOptions;
import io.kuzzle.sdk.Protocol.WebSocket;

import java.util.concurrent.ConcurrentHashMap;

public class updateDocument {
    private static Kuzzle kuzzle;

    public static void main(String[] args) {
        WebSocketOptions opts = new WebSocketOptions();
        opts.setAutoReconnect(true).setConnectionTimeout(42000);

        try {
            WebSocket ws = new WebSocket("localhost", opts);

            kuzzle = new Kuzzle(ws, (KuzzleOptions) null);

            kuzzle.connect();

            ConcurrentHashMap<String, Object> content = new ConcurrentHashMap<>();
            content.put("name", "not-so-random");

            ConcurrentHashMap<String, Object> options = new ConcurrentHashMap<>();
            options.put("source", true);

            ConcurrentHashMap<String, Object> response =
            kuzzle.getDocumentController().update("nyc-open-data", "yellow-taxi", "some-id", content, options)
            .get();
            System.out.println("Document updated!");
            System.out.println(response);
	}  catch (Exception e) {
            e.printStackTrace();
        }

        kuzzle.disconnect();
    }
};
```

You should see the document updated in your collection.
You can test without options as well

### Other changes

Update subscribe documentation -> Remove non existing snippets
Remove class `DocumentOptions` and add a `CreateOptions` class

### CODE REVIEW

File: doc/3/controllers/document/update/snippets/update.test.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: document#update
+description: Update a document
+hooks:
+  before: |
+    curl -XDELETE kuzzle:7512/nyc-open-data
+    curl -XPOST kuzzle:7512/nyc-open-data/_create
+    curl -XPUT kuzzle:7512/nyc-open-data/yellow-taxi
+  after:
+template: default
+expected: "id: 'some-id'"
```
Comment (scottinet) [inline doc/3/controllers/document/update/snippets/update.test.yml]: I'm not sure that this test can pass. Can you add documentation snippet tests to travis to make sure of that?

File: doc/3/controllers/document/update/index.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      final ConcurrentHashMap<String, Object> options)
+throws NotConnectedException, InternalException
+```
+
+| Arguments          | Type                                         | Description                       |
+| ------------------ | -------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                            | Index                             |
+| `collection`       | <pre>String</pre>                            | Collection                        |
+| `id        `       | <pre>String</pre>                            | Document ID                        |
+| `document`         | <pre>ConcurrentHashMap<String, Object></pre> | Partial content of the document to update |
```
Comment (Aschen) [inline doc/3/controllers/document/update/index.md]: Nitpicking
```suggestion
| `document`         | <pre>ConcurrentHashMap<String, Object></pre> | Partial document content |
```

File: doc/3/controllers/document/update/snippets/update.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+    ConcurrentHashMap<String, Object> content = new ConcurrentHashMap<>();
+    content.put("name", "Johny");
+
+    ConcurrentHashMap<String, Object> response =
+    kuzzle.getDocumentController().update("nyc-open-data", "yellow-taxi", "some-id", content)
+    .get();
+    System.out.println(response);
```
Comment (Aschen) [inline doc/3/controllers/document/update/snippets/update.java]: Can you add the result of the `println` in comment? 
It helps developer to better apprehend the structure of the response object they have. We try to do this for every method (Look the [Javascript SDK](https://github.com/kuzzleio/sdk-javascript/blob/master/doc/7/controllers/document/create/snippets/create.js) and [Doc](https://docs.kuzzle.io/sdk/js/7/controllers/document/create/#usage))

File: src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    query
+        .put("index", index)
+        .put("collection", collection)
+        .put("controller", "document")
+        .put("action", "update")
+        .put("body", document)
+        .put("_id",  id)
+        .put("waitForRefresh", _options != null ? _options.getBoolean("waitForRefresh") : null)
+        .put("retryOnConflict", _options != null ? _options.getNumber("retryOnConflict") : 0)
```
Comment (Aschen) [inline src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java]: You should let Kuzzle handle default values
```suggestion
        .put("retryOnConflict", _options != null ? _options.getNumber("retryOnConflict") : null)
```

Comment (Yoann-Abbes) [reply src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java]: When its null Kuzzle throz qn error

File: doc/3/controllers/document/update/index.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      final String id,
+      final ConcurrentHashMap<String, Object> document)
+throws NotConnectedException, InternalException
+
+public CompletableFuture<ConcurrentHashMap<String, Object>> update(
+      final String index,
+      final String collection,
+      final String id,
+      final ConcurrentHashMap<String, Object> document,
+      final ConcurrentHashMap<String, Object> options)
```
Comment (jenow) [inline doc/3/controllers/document/update/index.md]: We should have a `DocumentOptions` class containing `waitForRefresh`, `retryOnConflict` and `source`

File: src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java (hunk 4 above)
Comment (scottinet) [reply src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java]: Confirmed, but this sounds like a bug in Kuzzle. So I fixed it: https://github.com/kuzzleio/kuzzle/pull/1579

Aschen's point still stands.

Comment (Yoann-Abbes) [reply src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java]: Okay thanks!
So unfortunately, until it's released, the snippet test will fail

Comment (scottinet) [reply src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java]: Ok I'm releasing kuzzle 2.1.1 then

Comment (scottinet) [reply src/main/java/io/kuzzle/sdk/API/Controllers/DocumentController.java]: @Yoann-Abbes > kuzzle 2.1.1 released

File: doc/3/controllers/document/update/index.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      final String id,
+      final ConcurrentHashMap<String, Object> document)
+throws NotConnectedException, InternalException
+
+public CompletableFuture<ConcurrentHashMap<String, Object>> update(
+      final String index,
+      final String collection,
+      final String id,
+      final ConcurrentHashMap<String, Object> document,
+      final DocumentOptions options)
```
Comment (scottinet) [inline doc/3/controllers/document/update/index.md]: I disagree with that name: that option class can only be passed to methods having a use for all its exposed options, namely waitForRefresh, retryOnConflict and source. These last 2 in particular are only relevant for update (and possibly mUpdate) methods.

You have to design 1 option class per "use" (in term of functionalities), meaning that all options must have meaning for a particular method. If you use a "DocumentOption" class for methods having no use for the retryOnConflict option for instance, then this can be hard for users to understand.

Rename this to `UpdateOptions`, `DocumentUpdateOptions`, or whatever is more fitting for this method.

File: doc/3/controllers/document/update/index.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+| `document`         | <pre>ConcurrentHashMap<String, Object></pre> | Partial document content |
+| `options`          | <pre>ConcurrentHashMap<String, Object></pre> | Optional parameters               |
+
+---
+
+### options
+
+A [DocumentOptions](/sdk/java/3/core-classes/document-options) object.
+
+The `update` method takes into account those following argument:
```
Comment (scottinet) [inline doc/3/controllers/document/update/index.md]:
```suggestion
The following options can be set:
```

File: doc/3/controllers/document/update/index.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+| Arguments          | Type                                         | Description                       |
+| ------------------ | -------------------------------------------- | --------------------------------- |           |
+| `waitForRefresh`   | <pre>Boolean</pre>                           | If set to `true`, Kuzzle will wait for the persistence layer to finish indexing|
+| `retryOnConflict`  | <pre>Integer</pre>                           | The number of times the database layer should retry in case of version conflict
+| `source`           | <pre>Boolean</pre>                           | If true, returns the updated document inside the response
```
Comment (scottinet) [inline doc/3/controllers/document/update/index.md]:
```suggestion
| `retryOnConflict`  | <pre>Integer</pre>                           | The number of times the database layer should retry in case of version conflict |
| `source`           | <pre>Boolean</pre>                           | If true, returns the updated document inside the response |
```

File: doc/3/controllers/realtime/subscribe/index.md [hunk 9]
```diff
@@ -52,10 +52,3 @@ _Subscription to document notifications with scope option_
 
 <<< ./snippets/document-notifications-leave-scope.java
 
-_Subscription to message notifications_
```
Comment (scottinet) [inline doc/3/controllers/realtime/subscribe/index.md]: Why this change?
I see nothing about it in the PR description, and I don't get why you deactivate those snippets.

File: doc/3/controllers/document/update/index.md (hunk 6 above)
Comment (scottinet) [reply doc/3/controllers/document/update/index.md]: I just saw that DocumentOptions already existed prior to this PR, and contains an `id` option for methods such as `document.create`. This makes my point even more relevant: there is no use for that id option, and this can even be **very** confusing for users, as the update method already takes an "id" argument.

File: doc/3/controllers/realtime/subscribe/index.md (hunk 9 above)
Comment (Yoann-Abbes) [reply doc/3/controllers/realtime/subscribe/index.md]: Because those snippets does not exists, I should add it in the PR description in deed

File: doc/3/controllers/document/update/index.md (hunk 6 above)
Comment (Yoann-Abbes) [reply doc/3/controllers/document/update/index.md]: Changed :+1:
